File: api/routes/rooms.js
// backend/routes/rooms.js

const express = require("express");
const router = express.Router();
const { ClerkExpressWithAuth } = require("@clerk/clerk-sdk-node");
const clerkAuth = ClerkExpressWithAuth();

const Message = require("../models/Message");
const Room = require('../models/Room');

// GET /api/rooms/my - extract unique roomIds the user is part of via messages
router.get("/my", clerkAuth, async (req, res) => {
  const myId = req.auth.userId;
  try {
    // Find all distinct roomIds the user is a member of
    const roomIds = await Message.distinct("roomId", { 
      roomId: { $exists: true },
      senderId: myId // optionally also filter receiverId: myId if needed
    });

    res.json({ roomIds });
  } catch (err) {
    console.error("Error fetching room IDs from messages:", err);
    res.status(500).json({ error: "Failed to fetch rooms" });
  }
});

// routes/rooms.js
router.get('/:id', async (req, res) => {
  const room = await Room.findById(req.params.id);
  if (!room) return res.status(404).json({ error: 'Room not found' });
  console.log('room',room);
  res.json({ roomName: room.name ,roomPic:room.roomPic});
});

module.exports = router;

File: api/routes/rooms.py
from fastapi import APIRouter, HTTPException, Depends
from fastapi_clerk_auth import ClerkConfig, ClerkHTTPBearer, HTTPAuthorizationCredentials
from motor.motor_asyncio import AsyncIOMotorClient
from bson import ObjectId
import os

router = APIRouter()

# MongoDB connection
client = AsyncIOMotorClient(os.getenv("MONGO_URI", "mongodb+srv://skyees:raj1234%23A@skyees.oabwff3.mongodb.net/skyees?retryWrites=true&w=majority"))
db = client["skyees"]
messages_collection = db["messages"]
rooms_collection = db["rooms"]

# ✅ Clerk config (validates JWTs from Authorization header)
clerk_config = ClerkConfig(
    jwks_url="https://clerk.skyees.com/.well-known/jwks.json"
)
clerk_auth_guard = ClerkHTTPBearer(config=clerk_config)

# ✅ GET /api/rooms/my
@router.get("/my")
async def get_my_rooms(credentials: HTTPAuthorizationCredentials = Depends(clerk_auth_guard)):
    try:
        clerk_id = credentials.decoded.get("sub")
        if not clerk_id:
            raise HTTPException(status_code=401, detail="Unauthorized: No Clerk userId")

        # Find all distinct roomIds the user is a member of
        roomIds = await messages_collection.distinct(
            "roomId",
            {"roomId": {"$exists": True}, "senderId": clerk_id}
        )
        return {"roomIds": roomIds}
    except Exception as e:
        print("Error fetching room IDs from messages:", e)
        raise HTTPException(status_code=500, detail="Failed to fetch rooms")

# ✅ GET /api/rooms/{id}
@router.get("/{id}")
async def get_room(id: str, credentials: HTTPAuthorizationCredentials = Depends(clerk_auth_guard)):
    try:
        clerk_id = credentials.decoded.get("sub")
        if not clerk_id:
            raise HTTPException(status_code=401, detail="Unauthorized: No Clerk userId")

        room = await rooms_collection.find_one({"_id": ObjectId(id)})
        if not room:
            raise HTTPException(status_code=404, detail="Room not found")

        print("room", room)
        return {"roomName": room.get("name"), "roomPic": room.get("roomPic")}
    except Exception as e:
        print("Error fetching room:", e)
        raise HTTPException(status_code=500, detail="Server error fetching room")